feat: MCP session reuse per workflow run

55Kamiryo opened the pull request:
> [!IMPORTANT]
>
> 1. Make sure you have read our [contribution guidelines](https://github.com/langgenius/dify/blob/main/CONTRIBUTING.md)
> 1. Ensure there is an associated issue and you have been assigned to it
> 1. Use the correct syntax to link this PR: `Fixes #<issue number>`.

## Summary
Fixes: https://github.com/langgenius/dify/issues/24224

Persist MCP client sessions per workflow_execution_id so multi-step flows (navigate → screenshot) reuse the same browser/context, fixing the blank screenshot issue caused by recreating a fresh context on every call.

### Background
- Each MCP tool call previously opened/closed a new session, losing browser state, causing blank screenshots after navigation, and incurring high init latency. McpSessionManager.acquire also had a race that could leak orphaned clients under concurrent acquires.

### Fix
- McpSessionManager/McpSessionRegistry cache clients per workflow run with idle timeout and cleanup.
- MCPClient/MCPClientWithAuthRetry expose reusable connect().

### Tests
- make lint (pass)
- make type-check (pass)
- uv run --project api --dev dev/pytest/pytest_unit_tests.sh (pass; only existing DeprecationWarnings)
- Manual E2E: Playwright MCP browser_install → browser_navigate → browser_take_screenshot stays in a single session and returns correct page captures.

### Note
- Agent App calls still use the ephemeral MCP path because they don’t carry a workflow_execution_id. 
- To avoid changing that flow in this PR, I left it as-is. If we want session reuse for Agent App as well, I’ll file a follow-up issue/PR.

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

## Screenshots

MCP tool chains keep browser state and avoid re-init cost; Playwright install → navigate → screenshot runs in one mcp-session-id, producing correct captures instead of blank images.

| Before | After |
|--------|-------|
| <img width="346" height="347" alt="image" src="https://github.com/user-attachments/assets/68ad159c-4530-4ec1-b044-13949453577b" /> | <img width="344" height="355" alt="image" src="https://github.com/user-attachments/assets/5b7c78f7-5cf2-4796-ace2-b8b9ee767af0" />|

## Checklist

- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [x] I understand that this PR may be closed in case there was no previous discussion or issues. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [x] I've updated the documentation accordingly.
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

55Kamiryo committed:
feat:MCP session reuse per workflow run

55Kamiryo committed:
feat:MCP session reuse per workflow run

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

55Kamiryo committed:
Add TTL cleanup to MCP session registry and tidy MCP tool import

55Kamiryo commented on api/core/mcp/session_manager.py:
```diff
@@ -0,0 +1,114 @@
+            manager = cls._registry.pop(workflow_execution_id, None)
+
+        if manager:
+            manager.cleanup()
```
I’ve addressed this: McpSessionRegistry now wraps managers with timestamps and opportunistically purges expired entries inside get_manager. The TTL is set to 20 minutes to align with the default `WORKFLOW_MAX_EXECUTION_TIME`, and I added a unit test to verify cleanup of stale managers

55Kamiryo commented on api/core/tools/mcp_tool/tool.py:
```diff
@@ -160,6 +162,32 @@ def invoke_remote_mcp_tool(self, tool_parameters: dict[str, Any]) -> CallToolRes
         try:
+            if workflow_execution_id:
+                # Workflow path: reuse MCP session scoped by workflow_execution_id
+                from core.mcp.session_manager import McpSessionRegistry
```
addressed: the `McpSessionRegistry` import in `api/core/tools/mcp_tool/tool.py` is now at the module top with the rest of the imports.

55Kamiryo committed:
align mcp auth super calls and fix session key url

55Kamiryo committed:
fix mcp auth super usage

55Kamiryo committed:
fix: prevent session manager leaks

55Kamiryo committed:
skip MCP cleanup in subgraphs

55Kamiryo committed:
fix: pause reasons error

55Kamiryo committed:
fix: mypy error

Nov1c444 commented on api/core/tools/__base/tool_runtime.py:
```diff
@@ -19,6 +19,7 @@ class ToolRuntime(BaseModel):
     credentials: dict[str, Any] = Field(default_factory=dict)
     credential_type: CredentialType = Field(default=CredentialType.API_KEY)
     runtime_parameters: dict[str, Any] = Field(default_factory=dict)
+    workflow_execution_id: str | None = None
```
The current approach of passing `workflow_execution_id` through `ToolRuntime` has a few design concerns:

1. **Pollutes the generic `ToolRuntime` class**: `workflow_execution_id` is only used by `MCPTool` for session reuse, but it's added to the generic `ToolRuntime` class that all tool types share. Other tools (BuiltinTool, ApiTool, WorkflowTool) don't use this field at all.

2. **Post-hoc assignment**: The field is set after `tool_runtime` is created (lines 80-82), rather than being passed during construction. This violates the "initialize at construction time" principle.

**Suggested refactor:**

Since only `MCPTool` needs this value, it should be an instance attribute of `MCPTool` itself:

```python
# mcp_tool/tool.py
class MCPTool(Tool):
    workflow_execution_id: str | None = None  # MCPTool-specific attribute
```

Then in `tool_node.py`:

```python
if isinstance(tool_runtime, MCPTool) and system_variables:
    tool_runtime.workflow_execution_id = system_variables.workflow_execution_id
```

This keeps the responsibility where it belongs and avoids polluting the shared `ToolRuntime` class with MCP-specific concerns.

Nov1c444 commented on api/core/tools/mcp_tool/tool.py:
```diff
@@ -141,6 +142,8 @@ def invoke_remote_mcp_tool(self, tool_parameters: dict[str, Any]) -> CallToolRes
         from extensions.ext_database import db
         from services.tools.mcp_tools_manage_service import MCPToolManageService
 
+        workflow_execution_id = getattr(self.runtime, "workflow_execution_id", None)
```
`getattr(self.runtime, "workflow_execution_id", None)` is used even though `self.runtime` is typed as `ToolRuntime` which already has this field. This suggests type uncertainty and is a code smell.

Nov1c444 commented on api/core/workflow/graph_engine/graph_engine.py:
```diff
@@ -356,6 +359,19 @@ def _stop_execution(self) -> None:
+        """Release any long-lived MCP sessions bound to this workflow run."""
+        try:
+            system_variables = self._graph_runtime_state.variable_pool.system_variables
+            workflow_execution_id = getattr(system_variables, "workflow_execution_id", None)
```
`system_variables` is typed as `SystemVariable` which already has `workflow_execution_id` attribute — use `system_variables.workflow_execution_id` directly instead of `getattr`.

```python
workflow_execution_id = system_variables.workflow_execution_id
```

Nov1c444 commented:
Sorry for the late review. Could you please merge the latest changes from main?

55Kamiryo committed:
refactor: relocate MCP session management variables and ensure persistence

55Kamiryo committed:
fix: tidy GraphEngine init params after rebase

55Kamiryo committed:
fix format

55Kamiryo committed:
fix: ignore import graph_engine -> mcp.session_manager

55Kamiryo commented on api/core/workflow/graph_engine/graph_engine.py:
```diff
@@ -64,6 +65,7 @@ def __init__(
         graph: Graph,
         graph_runtime_state: GraphRuntimeState,
         command_channel: CommandChannel,
+        cleanup_mcp_sessions: bool = True,
```
Agreed. 

By moving cleanup_mcp_sessions to GraphRuntimeState and ensuring it is persisted in the snapshot, we achieved clarity and robustness across pause/resume cycles.

I briefly considered extracting this logic into a separate layer to decouple the dependencies. However, I concluded that it would be over-engineering at this stage, as it introduces unnecessary complexity and makes the session control in sub-graphs (like iterations) implicit and harder to track. Keeping it as an explicit flag in the state is the most pragmatic and readable approach.

55Kamiryo commented:
> Sorry for the late review. Could you please merge the latest changes from main?

@Nov1c444 
Thank you for the review. I've rebased on latest main and addressed your feedback.          

Key Decisions:
1. Explicit State Management: Moved cleanup_mcp_sessions to GraphRuntimeState (with persistence support).
2. Type Safety: Removed workflow_execution_id from the generic runtime and enforced strict typing with direct attribute access.
3. Persistence: Ensured that the cleanup flag is preserved in the database when the workflow is paused (e.g., for Human-in-the-loop), preventing session leaks upon resumption.

Verification:
- Unit tests passed (including pause/resume scenarios).
- Type checks passed.
- Verified integration with Playwright MCP.

Note: 
I'm aware that Agent Nodes also need similar MCP session handling, but I'll address that in a separate PR or issue since the implementation details differ.

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

55Kamiryo committed:
Merge branch 'main' into feat/mcp-session-manager

55Kamiryo committed:
Merge branch 'main' into feat/mcp-session-manager

QuantumGhost commented:
Hi, @55Kamiryo, sorry for slow reply. We have been refactoring the workflow recently. Could you please merge main again?

crazywoola commented:
Hi @55Kamiryo, thank you for taking the time to contribute to Dify.

### Why this is being closed

This PR was opened before February 2026 (`2025-11-25`) and touches more than 3 files (`12` files). Older broad PRs are very likely to conflict with the current codebase and are difficult to review safely after the surrounding code has moved on. Per our triage rules, we are closing old/conflicting PRs of this size to keep the review queue manageable.

### Next steps

If this change is still needed, please open a fresh PR from the latest `main`, keep the scope small, and link the related issue using GitHub's closing syntax such as `Fixes #123`, `Closes #123`, or `Resolves #123`. Thanks again for your work and understanding.
